File: packages/MeeshySDK/Sources/MeeshySDK/Models/AchievementCatalog.swift
import Foundation

/// LA GRAMMAIRE DES SUCCÈS, côté iOS (#5758/#5759) — MIROIR de
/// `packages/shared/types/achievement-catalog.ts` et `utils/achievement-view.ts`.
///
/// Le catalogue Xcode ne peut pas importer un module TypeScript : ce fichier en
/// est la projection, et il doit rendre EXACTEMENT les mêmes entrées, dans le
/// même ordre, que le web. Trois mécanismes distincts s'y rejouent, et les
/// confondre les casserait :
///
///  1. l'ATTEIGNABILITÉ, qui RETIRE un palier que le produit ne peut pas rendre
///     vrai — un objectif hors de portée ne doit pas exister, même invisible ;
///  2. l'ORDRE, du moins complexe au plus complexe, par `baseDifficulty +
///     log10(palier)` — le palier brut rangerait « 100 conversations » et
///     « une conversation de 100 membres » au même endroit ;
///  3. la FENÊTRE `max(7, acquis + 2)`, qui garde le prochain objectif visible
///     à tout moment.
public struct AchievementFamily: Sendable, Equatable, Hashable {
    public let section: String
    public let subject: String
    public let verb: String
    /// « count » (volume — combien de fois) ou « size » (ampleur — de quelle taille).
    public let scale: String
    public let baseDifficulty: Double

    public init(section: String, subject: String, verb: String, scale: String, baseDifficulty: Double) {
        self.section = section
        self.subject = subject
        self.verb = verb
        self.scale = scale
        self.baseDifficulty = baseDifficulty
    }

    /// L'identité dans les tables de libellés — sans la section, qui n'en change pas le sens.
    public var id: String { "\(subject).\(verb).\(scale)" }

    public func key(tier: Int) -> String {
        "achievement.\(section).\(subject).\(verb).\(scale):\(tier)"
    }

    public var tiers: [Int] {
        scale == "count" ? AchievementCatalog.countTiers : AchievementCatalog.sizeTiers
    }

    public func difficulty(tier: Int) -> Double {
        baseDifficulty + log10(Double(max(1, tier)))
    }
}

public enum AchievementCatalog {
    public static let countTiers: [Int] = [1, 10, 100, 1_000, 10_000]
    public static let sizeTiers: [Int] = [10, 100, 1_000, 10_000, 100_000, 1_000_000]

    /// La fenêtre : sept entrées au minimum, puis deux de plus par succès décroché.
    public static let windowMinimum = 7
    public static let windowStep = 2

    /// Les familles, TOUTES sections — miroir exact du TypeScript, gardé par
    /// `achievement-catalog-mirror-parity`.
    ///
    /// `decouverte` est ABSENTE des deux côtés : elle demande la forme
    /// COLLECTION (#5751), dont aucun stockage n'existe encore.
    ///
    ///
    /// `community.leave.count` est ABSENTE, comme côté TypeScript : quitter une
    /// communauté SUPPRIME la ligne, il n'y a rien à compter (#5760). La
    /// déclarer ici produirait un succès que rien ne peut faire tomber — et une
    /// divergence avec le web, que le témoin de parité interdit.
    public static let families: [AchievementFamily] = [
        AchievementFamily(section: "cercles", subject: "conversation", verb: "join", scale: "size", baseDifficulty: 1),
        AchievementFamily(section: "cercles", subject: "conversation", verb: "join", scale: "count", baseDifficulty: 2),
        AchievementFamily(section: "cercles", subject: "conversation", verb: "leave", scale: "count", baseDifficulty: 1.8),
        AchievementFamily(section: "cercles", subject: "conversation", verb: "create", scale: "count", baseDifficulty: 2.5),
        AchievementFamily(section: "cercles", subject: "community", verb: "join", scale: "size", baseDifficulty: 1.5),
        AchievementFamily(section: "cercles", subject: "community", verb: "join", scale: "count", baseDifficulty: 2.5),
        AchievementFamily(section: "cercles", subject: "community", verb: "create", scale: "size", baseDifficulty: 2),
        AchievementFamily(section: "cercles", subject: "community", verb: "create", scale: "count", baseDifficulty: 3),
        AchievementFamily(section: "parole", subject: "message", verb: "send", scale: "count", baseDifficulty: 0),
        AchievementFamily(section: "parole", subject: "voice", verb: "send", scale: "count", baseDifficulty: 0.8),
        AchievementFamily(section: "parole", subject: "image", verb: "send", scale: "count", baseDifficulty: 0.8),
        AchievementFamily(section: "parole", subject: "video", verb: "send", scale: "count", baseDifficulty: 1.3),
        AchievementFamily(section: "retouche", subject: "message", verb: "edit", scale: "count", baseDifficulty: 1),
        AchievementFamily(section: "retouche", subject: "message", verb: "delete", scale: "count", baseDifficulty: 1),
        AchievementFamily(section: "retouche", subject: "message", verb: "react", scale: "count", baseDifficulty: 0.5),
        AchievementFamily(section: "appels", subject: "call", verb: "join", scale: "count", baseDifficulty: 1.5),
        AchievementFamily(section: "appels", subject: "call", verb: "start", scale: "count", baseDifficulty: 2),
        AchievementFamily(section: "appels", subject: "call", verb: "start", scale: "size", baseDifficulty: 2.5),
        AchievementFamily(section: "ambassade", subject: "referral", verb: "complete", scale: "count", baseDifficulty: 2.5),
        AchievementFamily(section: "ambassade", subject: "link", verb: "click", scale: "count", baseDifficulty: 1.5),
        AchievementFamily(section: "constance", subject: "streak", verb: "hold", scale: "count", baseDifficulty: 2),
        AchievementFamily(section: "monnaie", subject: "meesh", verb: "mint", scale: "count", baseDifficulty: 3.5),
    ]

    /// Une famille absente de la carte n'est pas « zéro » : c'est « non mesuré ».
    /// Ses paliers d'AMPLEUR sont masqués (on ne promet pas ce qu'on ignore) et
    /// ses paliers de VOLUME restent visibles (répéter reste possible).
    public static func isAttainable(_ family: AchievementFamily, tier: Int, reach: [String: Int]) -> Bool {
        guard let mesure = reach[family.id] else { return family.scale == "count" }
        return tier <= mesure
    }

    /// **L'INVERSE de `AchievementFamily.key(tier:)`** — lire une clé composée
    /// telle que la passerelle la grave et la pousse (#5847).
    ///
    /// `achievement.<section>.<sujet>.<geste>.<échelle>:<palier>`. Une clé qui
    /// ne se range dans AUCUNE famille déclarée, ou dont le palier n'appartient
    /// pas à l'échelle de sa famille, rend `nil` : le catalogue ne l'a jamais
    /// produite, donc célébrer « une conversation de 42 membres » afficherait un
    /// badge bien dessiné pour un fait inventé. C'est le repli MENTEUR, le plus
    /// cher parce qu'il ne se voit qu'à la lecture.
    public static func parse(key: String) -> (family: AchievementFamily, tier: Int)? {
        let morceaux = key.split(separator: ":", omittingEmptySubsequences: false)
        guard morceaux.count == 2, let palier = Int(morceaux[1]) else { return nil }

        let chemin = morceaux[0].split(separator: ".", omittingEmptySubsequences: false)
        guard chemin.count == 5, chemin[0] == "achievement" else { return nil }

        guard let famille = families.first(where: {
            $0.section == chemin[1] && $0.subject == chemin[2]
                && $0.verb == chemin[3] && $0.scale == chemin[4]
        }) else { return nil }

        guard famille.tiers.contains(palier) else { return nil }
        return (famille, palier)
    }

    public static func windowSize(unlockedCount: Int) -> Int {
        max(windowMinimum, max(0, unlockedCount) + windowStep)
    }
}

public struct AchievementEntry: Sendable, Equatable, Identifiable {
    public let key: String
    public let section: String
    public let family: AchievementFamily
    public let tier: Int
    public let difficulty: Double
    public let unlocked: Bool
    public let reachedAt: Date?

    public var id: String { key }
}

public struct AchievementSectionView: Sendable, Equatable, Identifiable {
    public let section: String
    /// Déjà tronquées à la fenêtre — la vue n'en retire aucune.
    public let entries: [AchievementEntry]
    public let unlockedCount: Int
    /// Le total ATTEIGNABLE, jamais le total déclaré.
    public let attainableCount: Int

    public var id: String { section }
}

public enum AchievementResolver {
    /// Développe, ordonne, filtre et tronque — dans cet ordre, comme la loi partagée.
    public static func sections(
        reach: [String: Int],
        unlocked: [String: Date?],
        families: [AchievementFamily] = AchievementCatalog.families
    ) -> [AchievementSectionView] {
        var entries: [AchievementEntry] = []
        for family in families {
            for tier in family.tiers where AchievementCatalog.isAttainable(family, tier: tier, reach: reach) {
                let key = family.key(tier: tier)
                let estAcquis = unlocked.keys.contains(key)
                entries.append(
                    AchievementEntry(
                        key: key,
                        section: family.section,
                        family: family,
                        tier: tier,
                        difficulty: family.difficulty(tier: tier),
                        unlocked: estAcquis,
                        reachedAt: unlocked[key] ?? nil
                    )
                )
            }
        }
        // À difficulté égale la clé départage : un ordre STABLE vaut mieux qu'un
        // ordre joli mais changeant d'un rendu à l'autre — et c'est aussi ce qui
        // garantit l'identité avec le web.
        entries.sort { $0.difficulty == $1.difficulty ? $0.key < $1.key : $0.difficulty < $1.difficulty }

        var parSection: [String: [AchievementEntry]] = [:]
        var ordre: [String] = []
        for entry in entries {
            if parSection[entry.section] == nil { ordre.append(entry.section) }
            parSection[entry.section, default: []].append(entry)
        }

        return ordre.compactMap { section in
            guard let toutes = parSection[section] else { return nil }
            let acquis = toutes.filter(\.unlocked).count
            return AchievementSectionView(
                section: section,
                entries: Array(toutes.prefix(AchievementCatalog.windowSize(unlockedCount: acquis))),
                unlockedCount: acquis,
                attainableCount: toutes.count
            )
        }
    }
}

File: packages/shared/types/achievement-catalog.ts
/**
 * LA GRAMMAIRE DES SUCCÈS (#5758) — comment produire des MILLIERS de succès
 * sans écrire des milliers de libellés.
 *
 * ## Le problème que ce module résout
 *
 * Le porteur demande des succès du type « tu as intégré une conversation de
 * plus de 10, 100, 1 000, 10 000, 100 000, 1 000 000 membres », « même chose
 * pour une communauté, pour le départ, pour un appel lancé, pour un message,
 * un vocal, une image, une vidéo, un supprimé, un édité, une réaction — avec
 * des compteurs encore ! ».
 *
 * Énumérer ce produit cartésien donnerait des milliers d'entrées × 8 langues.
 * Ingérable, et faux dès la première divergence de traduction.
 *
 * **La clé est donc COMPOSÉE, et le libellé aussi** :
 *
 *     achievement.<section>.<sujet>.<geste>.<échelle>:<palier>
 *     achievement.cercles.conversation.join.size:1000
 *     achievement.parole.message.send.count:10000
 *
 * Le nombre de chaînes à traduire suit le nombre de COUPLES (sujet, geste,
 * échelle), pas le nombre de succès : un gabarit sert tous ses paliers.
 *
 * ## Deux échelles, qu'il ne faut jamais confondre
 *
 *  - **`size` — l'AMPLEUR** : la taille de ce qu'on a rejoint. UN seul
 *    événement peut débloquer un palier élevé (entrer dans un groupe de
 *    100 000 le donne d'un coup). Récompense la chance et l'audience.
 *  - **`count` — le VOLUME** : combien de fois. Exige la répétition.
 *    Récompense la constance.
 *
 * Les mélanger dans une même échelle produirait un classement absurde : « 100
 * conversations rejointes » et « une conversation de 100 membres » portent le
 * même nombre et n'ont pas la même difficulté du tout. D'où `baseDifficulty`,
 * déclaré par famille, qui rend les deux comparables.
 *
 * ## Ce que le catalogue REFUSE
 *
 *  - **Un palier inatteignable ne s'affiche pas.** Un succès « conversation de
 *    1 000 000 membres » dans un produit dont la plus grande en compte 300 est
 *    une promesse qu'on ne peut pas tenir. L'atteignabilité est MESURÉE sur la
 *    réalité du produit (`AchievementReach`), jamais déclarée une fois pour
 *    toutes.
 *  - **Un succès ne crédite AUCUN point.** Comme les records et les
 *    collections (#5751) : sinon la course aux succès deviendrait une pompe à
 *    Meeshes.
 */

/** Les SECTIONS — une rangée horizontale chacune dans l'écran Progression. */
export const ACHIEVEMENT_SECTIONS = [
  'cercles',
  'parole',
  'retouche',
  'appels',
  'ambassade',
  'constance',
  'monnaie',
  'decouverte',
] as const;

export type AchievementSection = (typeof ACHIEVEMENT_SECTIONS)[number];

/** L'échelle sur laquelle un palier se mesure. */
export const ACHIEVEMENT_SCALES = ['count', 'size'] as const;
export type AchievementScale = (typeof ACHIEVEMENT_SCALES)[number];

/** Paliers de VOLUME — « combien de fois ». */
export const ACHIEVEMENT_COUNT_TIERS = [1, 10, 100, 1_000, 10_000] as const;

/** Paliers d'AMPLEUR — « de quelle taille », tels que le porteur les a donnés. */
export const ACHIEVEMENT_SIZE_TIERS = [10, 100, 1_000, 10_000, 100_000, 1_000_000] as const;

/**
 * Une FAMILLE : un couple (sujet, geste) sur une échelle. C'est l'unité qui
 * porte un gabarit de libellé — donc l'unité qu'il faut traduire, une fois,
 * pour tous ses paliers.
 */
export type AchievementFamily = {
  readonly section: AchievementSection;
  /** Ce sur quoi porte le geste : `conversation`, `community`, `message`, `call`… */
  readonly subject: string;
  /** Le geste : `join`, `leave`, `start`, `send`, `edit`, `delete`, `react`… */
  readonly verb: string;
  readonly scale: AchievementScale;
  /**
   * Le coût de base de cette famille, qui rend les échelles COMPARABLES.
   * Rejoindre un groupe de 1 000 membres n'est pas aussi dur que rejoindre
   * 1 000 groupes : sans ce décalage, le tri par palier mentirait.
   */
  readonly baseDifficulty: number;
};

/** La clé stable d'un succès — la seule chose qui voyage et se grave. */
export const achievementKey = (family: AchievementFamily, tier: number): string =>
  `achievement.${family.section}.${family.subject}.${family.verb}.${family.scale}:${tier}`;

/** Les paliers déclarés d'une famille, dans l'ordre croissant. */
export const tiersOf = (family: AchievementFamily): readonly number[] =>
  family.scale === 'count' ? ACHIEVEMENT_COUNT_TIERS : ACHIEVEMENT_SIZE_TIERS;

/**
 * La DIFFICULTÉ d'un palier — le scalaire qui range le catalogue « du moins
 * complexe au plus complexe » (directive porteur).
 *
 * `log10` et non le palier brut : les paliers montent par puissances de dix,
 * donc leur logarithme est le nombre de CRANS, ce qui est la vraie mesure de
 * l'effort. Le tri devient alors comparable entre familles.
 */
export const difficultyOf = (family: AchievementFamily, tier: number): number =>
  family.baseDifficulty + Math.log10(Math.max(1, tier));

/**
 * Ce que le produit rend ATTEIGNABLE, mesuré — jamais supposé.
 *
 * Pour une échelle `size`, c'est la plus grande entité qui existe (la plus
 * grosse conversation, la plus grosse communauté). Pour `count`, c'est un
 * plafond de bon sens ou l'absence de plafond.
 *
 * Une famille absente de la carte est traitée comme NON MESURÉE : ses paliers
 * `size` sont masqués (on ne promet pas ce qu'on ne sait pas tenir) et ses
 * paliers `count` restent visibles (rien n'empêche structurellement de
 * répéter un geste).
 */
export type AchievementReach = ReadonlyMap<string, number>;

/** La clé de mesure d'une famille dans `AchievementReach`. */
export const reachKey = (family: AchievementFamily): string =>
  `${family.subject}.${family.verb}.${family.scale}`;
